Requirements.md

```diff
@@ -0,0 +1,26 @@
+## Requirements
+
+1. Lazy initialization.
+   - DBConfigurations.cs line: 10.  
+2. Generics (in delegates, events and methods)
+    - All ```Base``` classes and interfaces. 
+3. Delegates.
+   - RestaurantService.cs line: 99
+   - FoodService.cs line: 95
+4. Events and their usage: standard and custom.
+   - ```RestaurantService``` publishes ```RestaurantRegistered``` event. ```EmailService``` subscribes to this event. When event is raised, email about successful registration is sent to restaurant's email adress.
+5. Exceptions and dealing with them in a meaningfull way.
+   - todo??? (Custom exceptions)
```

Comment:
Good points of catching custom exceptions:
- Add semantic, BusinessLogicException, DatabaseException, etc is easier to read and understand than just Exception.
- We catch what we want. Imagine that we forgot a try-catch and we are catching Exception in the API, we could catch DatabaseException, BusinessLogicException, etc so that our responses could be misleading.  For example, 
```csharp
try
{
   RegisterUser(...);
} catch( BusinessLogicException e )
{
  SendReponse(BadRequest, "The user already exists");
}


try
{
   RegisterUser(...);
} catch( Exception e ).         //Maybe are catching other exception so this response does not make sense
{
  SendReponse(BadRequest, "The user already exists");
}
```

Reply:
I think this would be a great addition and shouldn't require much work. We would be certain that we meet the requirement.

Reply:
It definitely does not require much work, just refactor the code. Here it is a post that explain how to do that, https://docs.microsoft.com/en-us/dotnet/standard/exceptions/how-to-create-user-defined-exceptions

Reply:
I will look into it and implement it later.